# K Nearest Neighbors

This is the seventh practical exercise of our course [Data Sciecne Bootcamp](https://open.hpi.de/courses/datascience2023).
In this exercise, you are going to practice and create your first knn model.

---
\
Here is a reminder about the technical limitations on this programming platform **OpenJupyter**:

#### Inactivity and Saving

- After 15 minutes of inactivity, the Python kernel shuts down, but the kernel is easy to restart if you continue your work within 60 minutes.
- After 60 minutes of inactivity the data is deleted to free up resources for other users.
- Therefore, please save the notebook locally (click "File" -> "Download") if you are pausing your work on the notebook for more than one hour. (We have put a reminder to save the notebook after each task.)
- If you have to restore your work from a downloaded notebook, press the Upload Button and select your locally saved `.ipynb` file (confirm to overwrite).

#### Grading

- The due date for this exercise is the **27th of June (11:55 pm / 23:55 UTC)**
- The grading process relies on the existing cell order.
- Therefore: Do not create new cells and do not delete existing cells in the notebook.
- If you need/want to create new cells in the notebook, you can add them after the grading code.
- If you need/want to have a draft notebook, use the JupyterLab_playground.ipynb notebook provided in this directory

### INTRO

In [ ]:
# # Run this cell first and only once
# from Ex7_Grader import *
# !python3 Ex7_Grader.pyc

In [1]:
# You don't have to adjust this cell.  
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn.neighbors

df = pd.read_csv("Datasets/ToyotaCorolla.csv")

### Have a look into the data here
Feel free to add multiple cells

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1427 entries, 0 to 1426
Data columns (total 36 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Id                1427 non-null   int64 
 1   Model             1427 non-null   object
 2   Price             1427 non-null   int64 
 3   Age               1427 non-null   int64 
 4   Mfg_Month         1427 non-null   int64 
 5   Mfg_Year          1427 non-null   int64 
 6   KM                1427 non-null   int64 
 7   Fuel_Type         1427 non-null   object
 8   HP                1427 non-null   int64 
 9   Met_Color         1427 non-null   int64 
 10  Automatic         1427 non-null   int64 
 11  cc                1427 non-null   int64 
 12  Doors             1427 non-null   int64 
 13  Cylinders         1427 non-null   int64 
 14  Gears             1427 non-null   int64 
 15  Quarterly_Tax     1427 non-null   int64 
 16  Weight            1427 non-null   int64 
 17  Mfr_Guarantee 

In [3]:
df.head()

Id                                           Model  Price  Age  Mfg_Month  \
0   1   TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500   23         10   
1   2   TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750   23         10   
2   3  ?TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950   24          9   
3   4   TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950   26          7   
4   5     TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750   30          3   

   Mfg_Year     KM Fuel_Type  HP  Met_Color  ...  Central_Lock  \
0      2002  46986    Diesel  90          1  ...             1   
1      2002  72937    Diesel  90          1  ...             1   
2      2002  41711    Diesel  90          1  ...             0   
3      2002  48000    Diesel  90          0  ...             0   
4      2002  38500    Diesel  90          0  ...             1   

   Powered_Windows  Power_Steering  Radio  Mistlamps  Sport_Model  \
0                1               1      0          0            0   
1                0               1      0          0            0   
2                0               1      0          0            0   
3                0               1      0          0            0   
4                1               1      0          1            0   

   Backseat_Divider  Metallic_Rim  Radio_cassette  Tow_Bar  
0                 1             0               0        0  
1                 1             0               0        0  
2                 1             0               0        0  
3                 1             0               0        0  
4                 1             0               0        0  

[5 rows x 36 columns]

In [ ]:
# no

Is there a (technical) need to **label encode** (some) of the data?

In [ ]:
# no

Use the following features to build a KNN model, that predicts the *Mfg_year*:
- Age
- Weight
- Kilometers driven
- Price

In [6]:
## Decide on features and assign X and y.
feature_columns = df[['Age', 'Weight', 'KM', 'Price', 'Mfg_Year']]
X = feature_columns.drop(columns=['Mfg_Year'])
y = feature_columns['Mfg_Year']

If you want to label encode, do it in the cell below.

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Your solution:


Split the data in a 80-20 ratio.
Keep the variable names (X_train etc.)
Keep the random_state of *123*.  
**This is crucial**.

In [7]:
from sklearn.model_selection import train_test_split

# Your solution:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 123)

Is there a need to **standardize** (some) of the data? 
Analyze the data and tick a conclusion as offered in the solution sheet.

If you want to standardize, do it in the cell below.

In [15]:
#Import library
from sklearn.preprocessing import StandardScaler

# Your solution:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Initiate the model with **15** neigbors, fit the model and validate the results with the test proportion of the input variables.

In [16]:
# Loading classifier libraries
from sklearn.neighbors import KNeighborsClassifier

# Your solution:


# Set neighbors that vote on a "new" entity to 15

classifier = KNeighborsClassifier(n_neighbors=15)

# Fitting the model

classifier.fit(X_train_scaled, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test_scaled)

[[ 0.2229408  -0.93229952  0.43946108 -0.1467719 ]
 [-0.20711068 -0.43650646 -0.51141596 -0.21905305]
 [-0.26086711  1.34834855  0.92115537 -0.01666584]
 ...
 [ 0.92177444 -0.7339823   0.01139532 -0.10687271]
 [-0.36837998  0.0592866   1.67895903 -0.30579042]
 [-0.04584138 -0.7339823  -0.85965382 -0.00365523]]


Check your model for accuracy, as shown in the video.

In [23]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

# Your solution:
print(classifier.score(X_test_scaled, y_test))
print(accuracy_score(y_test, y_pred))

0.8881118881118881
0.8881118881118881


Make sure to know the accuracy written in rounded percentage:  
0.11 would be 11%

## Draft cells